src/exoticatechnologies/modifications/exotics/impl/RewindSystem.kt

```diff
@@ -0,0 +1,488 @@
+import org.magiclib.subsystems.MagicSubsystemsManager
+import java.awt.Color
+import java.util.concurrent.atomic.AtomicBoolean
+
+class RewindSystem(key: String, settings: JSONObject) : Exotic(key, settings) {
+    override var color = Color(225, 225, 225)
+    private lateinit var originalShip: ShipAPI
+
+    private val logger: Logger = Logger.getLogger(RewindSystem::class.java)
+
+    override fun canAfford(fleet: CampaignFleetAPI, market: MarketAPI?): Boolean {
+        return Utilities.hasItem(fleet.cargo, ITEM)
+    }
+
+    override fun removeItemsFromFleet(fleet: CampaignFleetAPI, member: FleetMemberAPI, market: MarketAPI?): Boolean {
+        Utilities.takeItemQuantity(fleet.cargo, ITEM, 1f)
+        return true
+    }
+
+
+    override fun modifyToolTip(
+            tooltip: TooltipMakerAPI,
+            title: UIComponentAPI,
+            member: FleetMemberAPI,
+            mods: ShipModifications,
+            exoticData: ExoticData,
+            expand: Boolean) {
+        StringUtils.getTranslation(key, "longDescription")
+                .format("how_much", calculateTooltipStringReplacement())
+                .format("cooldown_string", cooldownString)
+                .addToTooltip(tooltip, title)
+    }
+
+    override fun applyToShip(
+            id: String,
+            member: FleetMemberAPI,
+            ship: ShipAPI,
+            mods: ShipModifications,
+            exoticData: ExoticData) {
+        super.applyToShip(id, member, ship, mods, exoticData)
+
+        originalShip = ship
+        val subsystem = RewindSubsystem(ship)
+        MagicSubsystemsManager.addSubsystemToShip(ship, subsystem)
+    }
+
+    inner class RewindSubsystem(ship: ShipAPI) : MagicSubsystem(ship) {
+        private var engine: CombatEngineAPI = Global.getCombatEngine()
+        private val secondsTracker = IntervalUtil(0.95f, 1.05f)
+        private val arcTimer = IntervalUtil(0.25f, 0.35f)//IntervalUtil(0.25f, 2.5f) //IntervalUtil(0.25f, 0.5f)
+        private val teleportTimer = IntervalUtil(PHASE_IN_DURATION - 0.05f, PHASE_IN_DURATION + 0.05f)//IntervalUtil(1.95f, 2.05f)
+        private var timeElapsed: Float = 0f
+        private val systemActivated = AtomicBoolean(false)
+        private val justTurnedOff = AtomicBoolean(false)
+        private val performedTeleport = AtomicBoolean(false)
+        /**
+        Measures *when* the system was activated and not *how long* it's been active for.
+        Calculate how long it's active for by doing *(timeElapsed - activationTime)*
+
+         If the value is less than 0, that should be interpreted as if it's not activated.
+         */
+        private var activationTime = -1f
+        private val previousStates: RingBuffer<ShipParams> = RingBuffer<ShipParams>(
+                determineRewindLength(ship) + 5,
+                ShipParams.EmptyShipParams,
+                ShipParams::class.java
+        )
+
+        //this should be either empty, or have an actual value
```

Comment:
useless comment is useless because it's a leftover from when a value-or-null approach was used. was a reminder to not use 0 or -1 or any other value to indicate "invalidness".

delete

Reply:
done